### ISSUE TITLE
migration doesn't work on my project

### ISSUE THREAD

**Issue by MahdiPishguy** [simolus3/moor#135]
simply i want to add table for my application database using `migration`. after make table schema and `dao` in database class i expected the table can be create on running project

what i doing?

1) make table class schema:

```dart
class UserBloodSugarTable extends Table{
  IntColumn get id              => integer().autoIncrement()();
  IntColumn get value           => integer()();
  DateTimeColumn get createdAt  => dateTime()();
}
```

2) make table dao and adding that into `tables`, `daos`:

```dart
@UseDao(tables:[UserBloodSugarTable])
class UserBloodSugarTableDao extends DatabaseAccessor<AppDb> with _$UserBloodSugarTableDaoMixin{
  final AppDb db;
  UserBloodSugarTableDao(this.db):super(db);

  Future<List<UserBloodSugarTableData>> getStoredUserBloodSugar() => select(userBloodSugarTable).get();
  Stream<List<UserBloodSugarTableData>> streamStoredUserBloodSugar() => select(userBloodSugarTable).watch();
  Future<int> insertBloodSugar(Insertable userBloodSugar) => into(userBloodSugarTable)  .insert(userBloodSugar);
  Future  updateBloodSugar(Insertable userBloodSugar) => update(userBloodSugarTable).replace(userBloodSugar);
  Future  deleteBloodSugar(Insertable userBloodSugar) => delete(userBloodSugarTable).delete(userBloodSugar);
}
```

UseMoor:

```dart
@UseMoor(
    tables:[
      ... ,
      UserBloodSugarTable
    ],
    daos: [
      ... , UserBloodSugarTableDao
    ],
    queries: {
    })
``` 

3) incrementing `schemaVersion`:

```dart
int get schemaVersion => 2;
```

4)implementing `migration`:

```dart
  @override
  MigrationStrategy get migration => MigrationStrategy(
      onUpgrade: (m, from, to) async {
        if (from <= 2) {
          await m.createTable(userBloodSugarTable);
        }
      },
    );
```

5) run build command:

```dart
flutter packages pub run build_runner build
```

6) close application and clear from device history:

7) build and run again application

```flutter run```

and then when i try to get query or insert data i get this error :

```dart
Unhandled Exception: DatabaseException(no such table: user_blood_sugar_table (Sqlite code 1): , while compiling: SELECT * FROM user_blood_sugar_table;, (OS error - 2:No such file or directory)) sql 'SELECT * FROM user_blood_sugar_table;' args []}
```

**Comment by simolus3** [comment]
Thank you for the detailed issue, I'll try to reproduce this. We've gotten reports for similar problems that only happened on some devices, so can you tell me what phone and which Android version you're running?

In the meantime, if you're not using type converters, can you try to use moor 1.6:
```yaml
dependencies:
  moor: '>=1.6.0 <1.7.0'
  moor_flutter: '>=1.6.0 <1.7.0'
dev_dependencies:
  moor_generator: '>=1.6.0 <1.7.0'
```
And let me know if the same problem happens on that version? That could very much help me finding the cause of this problem.

**Comment by simolus3** [comment]
I tried to follow your steps as closely as possible, but I still couldn't reproduce the problem. I have identified a potential cause, but it's hard to tell for sure without reproducing this. Can you put this in your pubspec:
```yaml
dependency_overrides:
  moor:
   git:
    url: https://github.com/simolus3/moor.git
    ref: fix-open
    path: moor/
  moor_flutter:
    git:
      url: https://github.com/simolus3/moor.git
      ref: fix-open
      path: moor_flutter/
```
And let me know if the problem persists or an assertion error is thrown?

**Comment by MahdiPishguy** [comment]
@simolus3 My android version is 8.0
>
>  can you tell me what phone and which Android version you're running?
> 

I get error after adding

```dart
dependency_overrides:
  moor:
   git:
    url: https://github.com/simolus3/moor.git
    ref: fix-open
    path: moor/
  moor_flutter:
    git:
      url: https://github.com/simolus3/moor.git
      ref: fix-open
      path: moor_flutter/
```

```
E/flutter (14134): [ERROR:flutter/shell/common/shell.cc(184)] Dart Error: Unhandled exception:
E/flutter (14134): 'package:moor_flutter/moor_flutter.dart': Failed assertion: line 57 pos 9: '!_debugIsOpening && !isOpen': Database opened multiple times, this should never happen. Please report this at https://github.com/simolus3/moor/issues/135 to help fix this!
E/flutter (14134): #0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:40:39)
E/flutter (14134): #1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
E/flutter (14134): #2      _SqfliteDelegate.open (package:moor_flutter/moor_flutter.dart:57:9)
E/flutter (14134): <asynchronous suspension>
E/flutter (14134): #3      DelegatedDatabase.ensureOpen.<anonymous closure> (package:moor/src/runtime/executor/helpers/engines.dart:247:22)
E/flutter (14134): <asynchronous suspension>
E/flutter (14134): #4      _DelegatedDatabase&QueryExecutor&_ExecutorWithQueryDelegate._synchronized (package:moor/src/runtime/executor/helpers/engines.dart:22:26)
E/flutter (14134): <asynchronous suspension>
E/flutter (14134): #5      DelegatedDatabase.ensureOpen (package:moor/src/runtime/executor/helpers/engines.dart:243:12)
E/flutter (14134): #6      QueryExecutor.doWhenOpened (package:moor/src/runtime/executor/executor.dart:26:12)
E/flutter (14134): #7      SimpleSelectStatement._getWithQuery (package:moor/src/runtime/statements/select.dart:191:40)
E/flutter (14134): <asynchronous suspension>
E/flutter (14134): #8      SimpleSelectStatement.get (package:moor/src/runtime/statements/select.dart:187:12)
E/flutter (14134): <asynchronous suspension>
E/flutter (14134): #9      _SimpleSelectStatement&Query&SingleTableQueryMixin&LimitContainerMixin&Selectable.getSingle (package:moor/src/runtime/statements/query.dart:98:24)
E/flutter (14134): <asynchronous suspension>
E/flutter (14134): #10     UserTableDao.getSingleUser (package:diabetist_pro/data/database/moor_database.dart:155:11)
E/flutter (14134): #11     ManageFragmentsRoutersState.build.<anonymous closure> (package:diabetist_pro/main.dart:242:68)
E/flutter (14134): #12     BlocBuilder.build (package:flutter_bloc/src/bloc_builder.dart:83:50)
E/flutter (14134): #13     _BlocBuilderBaseState.build (package:flutter_bloc/src/bloc_builder.dart:150:48)
E/flutter (14134): #14     StatefulElement.build (package:flutter/src/widgets/framework.dart:4012:27)
E/flutter (14134): #15     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3924:15)
E/flutter (14134): #16     Element.rebuild (package:flutter/src/widgets/framework.dart:3721:5)
E/flutter (14134): #17     StatefulElement.update (package:flutter/src/widgets/framework.dart:4085:5)
E/flutter (14134): #18     Element.updateChild (package:flutter/src/widgets/framework.dart:2876:15)
E/flutter (14134): #19     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3935:16)
E/flutter (14134): #20     Element.rebuild (package:flutter/src/widgets/framework.dart:3721:5)
E/flutter (14134): #21     StatefulElement.update (package:flutter/src/widgets/framework.dart:4085:5)
E/flutter (14134): #22     Element.updateChild (package:flutter/src/widgets/framework.dart:2876:15)
E/flutter (14134): #23     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3935:16)
E/flutter (14134): #24     Element.rebuild (package:flutter/src/widgets/framework.dart:3721:5)
E/flutter (14134): #25     StatelessElement.update (package:flutter/src/widgets/framework.dart:3981:5)
E/flutter (14134): #26     Element.updateChild (package:flutter/src/widgets/framework.dart:2876:15)
E/flutter (14134): #27     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3935:16)
E/flutter (14134): #28     Element.rebuild (package:flutter/src/widgets/framework.dart:3721:5)
E/flutter (14134): #29     StatefulElement.update (package:flutter/src/widgets/framework.dart:4085:5)
E/flutter (14134): #30     Element.updateChild (package:flutter/src/widgets/framework.dart:2876:15)
E/flutter (14134): #31     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3935:16)
E/flutter (14134): #32     Element.rebuild (package:flutter/src/widgets/framework.dart:3721:5)
E/flutter (14134): #33     ProxyElement.update (package:flutter/src/widgets/framework.dart:4219:5)
E/flutter (14134): #34     Element.updateChild (package:flutter/src/widgets/framework.dart:2876:15)
E/flutter (14134): #35     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3935:16)
E/flutter (14134): #36     Element.rebuild (package:flutter/src/widgets/framework.dart:3721:5)
E/flutter (14134): #37     StatefulElement.update (package:flutter/src/widgets/framework.dart:4085:5)
E/flutter (14134): #38     Element.updateChild (package:flutter/src/widgets/framework.dart:2876:15)
E/flutter (14134): #39     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5099:14)
```

**Comment by simolus3** [comment]
Thank you, that actually helped me a lot! I was able to reproduce this issue by adding a delay when opening the database. It should be fixed now, can you try to run `flutter packages upgrade`? If it uses moor at commit 3c56e0a and prints "uses fix for the opening problem" at the start of your app, should have the version with the patches. Can you verify whether that works for you?

**Comment by MahdiPishguy** [comment]
> 
> 
> Thank you, that actually helped me a lot! I was able to reproduce this issue by adding a delay when opening the database. It should be fixed now, can you try to run `flutter packages upgrade`? If it uses moor at commit [3c56e0a](https://github.com/simolus3/moor/commit/3c56e0ad6e2598a5c394213c92a65273b7420a1c) and prints "uses fix for the opening problem" at the start of your app, should have the version with the patches. Can you verify whether that works for you?

I get error again:

```dart
E:\Projects\Flutter\diabetist>flutter packages upgrade
! moor 1.7.1 from git https://github.com/simolus3/moor.git at 3c56e0 in moor/
! moor_flutter 1.7.0 from git https://github.com/simolus3/moor.git at 3c56e0 in moor_flutter/
Running "flutter pub upgrade" in diabetist...                      55.1s
```

and console output

```
E/flutter (23580): [ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: 'package:moor_flutter/moor_flutter.dart': Failed assertion: line 57 pos 9: '!_debugIsOpening && !isOpen': Database opened multiple times, this should never happen. Please report this at https://github.com/simolus3/moor/issues/135 to help fix this!
E/flutter (23580): #0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:40:39)
E/flutter (23580): #1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
E/flutter (23580): #2      _SqfliteDelegate.open (package:moor_flutter/moor_flutter.dart:57:9)
E/flutter (23580): <asynchronous suspension>
E/flutter (23580): #3      DelegatedDatabase.ensureOpen.<anonymous closure> (package:moor/src/runtime/executor/helpers/engines.dart:247:22)
E/flutter (23580): <asynchronous suspension>
E/flutter (23580): #4      _DelegatedDatabase&QueryExecutor&_ExecutorWithQueryDelegate._synchronized (package:moor/src/runtime/executor/helpers/engines.dart:22:26)
E/flutter (23580): <asynchronous suspension>
E/flutter (23580): #5      DelegatedDatabase.ensureOpen (package:moor/src/runtime/executor/helpers/engines.dart:243:12)
E/flutter (23580): #6      QueryExecutor.doWhenOpened (package:moor/src/runtime/executor/executor.dart:26:12)
E/flutter (23580): #7      SimpleSelectStatement._getWithQuery (package:moor/src/runtime/statements/select.dart:191:40)
E/flutter (23580): <asynchronous suspension>
E/flutter (23580): #8      SimpleSelectStatement.get (package:moor/src/runtime/statements/select.dart:187:12)
E/flutter (23580): <asynchronous suspension>
E/flutter (23580): #9      _SimpleSelectStatement&Query&SingleTableQueryMixin&LimitContainerMixin&Selectable.getSingle (package:moor/src/runtime/statements/query.dart:98:24)
E/flutter (23580): <asynchronous suspension>
E/flutter (23580): #10     UserTableDao.getSingleUser (package:diabetist_pro/data/database/moor_database.dart:155:11)
E/flutter (23580): #11     ManageFragmentsRoutersState.build.<anonymous closure> (package:diabetist_pro/main.dart:242:68)
E/flutter (23580): #12     BlocBuilder.build (package:flutter_bloc/src/bloc_builder.dart:83:50)
E/flutter (23580): #13     _BlocBuilderBaseState.build (package:flutter_bloc/src/bloc_builder.dart:150:48)
E/flutter (23580): #14     StatefulElement.build (package:flutter/src/widgets/framework.dart:4012:27)
E/flutter (23580): #15     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3924:15)
E/flutter (23580): #16     Element.rebuild (package:flutter/src/widgets/framework.dart:3721:5)
E/flutter (23580): #17     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2340:33)
E/flutter (23580): #18     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:700:20)
E/flutter (23580): #19     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:285:5)
E/flutter (23580): #20     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1016:15)
E/flutter (23580): #21     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:958:9)
E/flutter (23580): #22     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:874:5)
E/flutter (23580): #23     _rootRun (dart:async/zone.dart:1124:13)
E/flutter (23580): #24     _CustomZone.run (dart:async/zone.dart:1021:19)
E/flutter (23580): #25     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
E/flutter (23580): #26     _invoke (dart:ui/hooks.dart:236:10)
E/flutter (23580): #27     _drawFrame (dart:ui/hooks.dart:194:3)

```

**Comment by simolus3** [comment]
Thanks for the feedback. I added some more diagnostics information at a17ad13
Can you try running `flutter clean` and `flutter packages upgrade` again? Before the error, there should be logs looking like `ensure_open_<num>: <message>`. Can you post these as well?

Sorry if I'm bothering you, this bug is pretty severe and I'd like to release a fix as quickly as possible :crossed_fingers:

**Comment by MahdiPishguy** [comment]
@simolus3 
> Sorry if I'm bothering you, this bug is pretty severe and I'd like to release a fix as quickly as possible 🤞

no matter, we want to solve this problem. i'm this library user and i like to help you

```dart
I/flutter (25942): ensure_open_1: called
I/flutter (25942): ensure_open_1: lock aquired
I/flutter (25942): ensure_open_2: called
I/flutter (25942): 5.0
I/flutter (25942): ensure_open_3: called
I/flutter (25942): 5.0
I/flutter (25942): ensure_open_1: opening database
I/flutter (25942): ensure_open_1: running migrations
I/flutter (25942): ensure_open_1: done with opening
I/flutter (25942): ensure_open_2: lock aquired
I/flutter (25942): Moor: Sent SELECT * FROM user_table ORDER BY id DESC LIMIT 1; with args []
I/flutter (25942): ensure_open_2: was already open
I/flutter (25942): ensure_open_3: lock aquired
I/flutter (25942): Moor: Sent SELECT * FROM user_table ORDER BY id DESC LIMIT 1; with args []
I/flutter (25942): ensure_open_3: was already open
I/flutter (25942): Moor: Sent SELECT * FROM user_table ORDER BY id DESC LIMIT 1; with args []
I/flutter (25942): manager navigateToName:-1
I/flutter (25942): manager navigateToName:1
I/flutter (25942): ensure_open_4: called
I/flutter (25942): ensure_open_4: lock aquired
I/flutter (25942): ensure_open_5: called
I/flutter (25942): 5.0
I/flutter (25942): ensure_open_6: called
I/flutter (25942): ensure_open_4: was already open
I/flutter (25942): ensure_open_5: lock aquired
I/flutter (25942): Moor: Sent SELECT * FROM user_table ORDER BY id DESC LIMIT 1; with args []
I/flutter (25942): ensure_open_5: was already open
I/flutter (25942): ensure_open_6: lock aquired
I/flutter (25942): Moor: Sent SELECT * FROM user_table ORDER BY id DESC LIMIT 1; with args []
I/flutter (25942): ensure_open_6: was already open
I/flutter (25942): Moor: Sent SELECT * FROM user_table ORDER BY id DESC LIMIT 1; with args []
I/zygote64(25942): Do partial code cache collection, code=30KB, data=27KB
I/zygote64(25942): After code cache collection, code=30KB, data=27KB
I/zygote64(25942): Increasing code cache capacity to 128KB
I/zygote64(25942): Do partial code cache collection, code=61KB, data=43KB
I/zygote64(25942): After code cache collection, code=61KB, data=43KB
I/zygote64(25942): Increasing code cache capacity to 256KB
I/flutter (25942): ensure_open_7: called
I/flutter (25942): ensure_open_7: lock aquired
I/flutter (25942): ensure_open_8: called
I/flutter (25942): 5.0
I/flutter (25942): ensure_open_9: called
I/flutter (25942): 5.0
I/flutter (25942): ensure_open_7: was already open
I/flutter (25942): ensure_open_8: lock aquired
I/flutter (25942): Moor: Sent SELECT * FROM user_table ORDER BY id DESC LIMIT 1; with args []
I/flutter (25942): ensure_open_8: was already open
I/flutter (25942): ensure_open_9: lock aquired
I/flutter (25942): Moor: Sent SELECT * FROM user_table ORDER BY id DESC LIMIT 1; with args []
I/flutter (25942): ensure_open_9: was already open
I/flutter (25942): Moor: Sent SELECT * FROM user_table ORDER BY id DESC LIMIT 1; with args []
```

It certainly makes me proud

**Comment by simolus3** [comment]
These logs look good, are you getting any errors? I've just published moor version 1.7.2 to pub which contains the fix, so you can "reset" your pubspec (remove the `dependency_overrides`) and run `flutter packages clean` and `flutter packages upgrade`.

Again, thanks a lot for your help!

**Comment by MahdiPishguy** [comment]
which them of your libraries? moor? moor_flutter? generation?

**Comment by simolus3** [comment]
I released moor `1.7.2`, the flutter and generator packages should stay the same. You can put an explicit `moor: >=1.7.2` in your pubspec to be safe, but I would expect `flutter packages upgrade` to pick up the transitive dependency either way.

**Comment by MahdiPishguy** [comment]
@simolus3 

for `moor: >=1.7.2` i get `invalid block scalar header` error and when i try to change this lines to latest version

```
  moor_flutter: ^1.7.2
  moor_generator: ^1.7.2
```

i get this error

```
Because diabetist_pro depends on moor_generator ^1.7.2 which doesn't match any versions, version solving failed.
pub get failed (1)
```

**Comment by simolus3** [comment]
Sorry if I was unclear, I'm using this pubspec to get the latest moor version:
```yaml

dependencies:
  moor_flutter: ^1.7.0
  moor: ">=1.7.2"
dev_dependencies:
  moor_generator: ^1.7.0
```

**Comment by MahdiPishguy** [comment]
@simolus3

i'm wondering why where you are sure this bug fixed 

```dart
E/flutter ( 2174): DatabaseException(no such table: user_blood_sugar_table (Sqlite code 1): , while compiling: SELECT * FROM user_blood_sugar_table;, (OS error - 2:No such file or directory)) sql 'SELECT * FROM user_blood_sugar_table;' args []}
```

that means problem doesn't work

**Comment by simolus3** [comment]
That's weird :/ I still couldn't reproduce the issue, even when I'm adding the fake delay when opening the database.

It might have already upgraded to `schemaVersion=2` with the broken moor version (which means the table wasn't created). Now, when you open the database it's at schema version 2 which means the table still isn't created. Can you try running `flutter clean` and then do this with the fixed version (1.7.2):
```dart
  int get schemaVersion => 3;

  @override
  MigrationStrategy get migration => MigrationStrategy(
      onUpgrade: (m, from, to) async {
        if (from <= 3) {
          await m.createTable(userBloodSugarTable);
        }
      },
    );
```

If that still doesn't work the problem is still present.

**Comment by MahdiPishguy** [comment]
@simolus3 

that's worked now, 😄 

That's weird, why in second version of `schemaVersion` it doesn't work.
